synapse/lib/storm.py

```diff
@@ -918,3 +925,79 @@ def _stormOperDelTag(self, query, oper):
 
         for tag in tags:
             [ core.delTufoTag(node,tag) for node in nodes ]
+
+    def _stormOperJoinTags(self, query, oper):
+        args = oper[1].get('args')
+        opts = dict(oper[1].get('kwlist'))
+
+        nodes = query.data()
+
+        keep_nodes = opts.get('keep_nodes', False)
+        limt = LimitHelp(opts.get('limit'))
+
+        forms = {arg for arg in args}
+
+        tags = {tag for node in nodes for tag in s_tufo.tags(node, leaf=True)}
+
+        core = self.getStormCore()
+
+        if forms:
+            tagforms = [(tag, tufo[1].get('syn:tagform:form')) for tag in tags
+                        for tufo in core.getTufosByProp('syn:tagform:tag', tag)
+                        if tufo[1].get('syn:tagform:form') in forms]
+        else:
+            tagforms = [(tag, tufo[1].get('syn:tagform:form')) for tag in tags
+                        for tufo in core.getTufosByProp('syn:tagform:tag', tag)]
+        tagforms = sorted(tagforms)
+        # This allows us to do 'join' source tufos together.
+        if not keep_nodes:
+            query.clear()
+
+        for tag, form in tagforms:
+            lqs = len(query.results.get('data'))
+            tufos = core.getTufosByTag(form, tag, limit=limt.get())
+            [query.add(tufo) for tufo in tufos]
+            if tufos:
+                lq = len(query.results.get('data'))
+                nnewnodes = lq - lqs
+                if limt.dec(nnewnodes):
+                    break
+
+    def _stormOperToTags(self, query, oper):
+        opts = dict(oper[1].get('kwlist'))
+        nodes = query.data()
+        core = self.getStormCore()
+
+        leaf = opts.get('leaf', True)
+        tags = {tag for node in nodes for tag in s_tufo.tags(node, leaf=leaf)}
+        query.clear()
+        [query.add(tufo) for tag in tags for tufo in core.getTufosByProp('syn:tag', tag)]
+
+    def _stormOperFromTags(self, query, oper):
+        args = oper[1].get('args')
+        opts = dict(oper[1].get('kwlist'))
+
+        nodes = query.data()
+
+        forms = {arg for arg in args}
+        limt = LimitHelp(opts.get('limit'))
+
+        tags = {node[1].get('syn:tag') for node in nodes if node[1].get('tufo:form') == 'syn:tag'}
+
+        core = self.getStormCore()
+        # TODO This is O(m*n).
+        if not forms:
+            forms = {tufo[1].get('syn:type') for tufo in core.getTufosByProp('tufo:form', 'syn:type')}
+        # Predictable looping
+        tagforms = sorted(itertools.product(tags, forms))
+
+        query.clear()
+        for tag, form in tagforms:
+            lqs = len(query.results.get('data'))
+            tufos = core.getTufosByTag(form, tag, limit=limt.get())
+            [query.add(tufo) for tufo in tufos]
+            if tufos:
+                lq = len(query.results.get('data'))
```

Comment:
this pierces into query object's locals.  Would it be better if we made a size() or __len__ impl on query object to answer this for us?